src/bun.js/node/types.zig

```diff
@@ -1840,548 +1861,2885 @@ pub const Emitter = struct {
+                };
+            } else {
+                // Translated from the following JS code:
+                //   path = process.env[`=${resolvedDevice}`] || process.cwd();
+                if (comptime Environment.isWindows) {
+                    // Windows has the concept of drive-specific current working
+                    // directories. If we've resolved a drive letter but not yet an
+                    // absolute path, get cwd for that drive, or the process cwd if
+                    // the drive cwd is not available. We're sure the device is not
+                    // a UNC path at this points, because UNC paths are always absolute.
+
+                    // Translated from the following JS code:
+                    //   process.env[`=${resolvedDevice}`]
+                    const key_w: [*:0]const u16 = brk: {
+                        if (resolvedDeviceLen == 2 and resolvedDevice[1] == CHAR_COLON) {
+                            // Fast path for device roots
+                            break :brk &[3:0]u16{ '=', resolvedDevice[0], CHAR_COLON };
+                        }
+                        bufSize = 1;
+                        // Reuse varedLenBuf for the env key because it's used to get the path.
+                        varedLenBuf[0] = '=';
+                        bufOffset = bufSize;
+                        bufSize += resolvedDeviceLen;
+                        @memcpy(varedLenBuf[bufOffset ..bufSize], resolvedDevice);
+                        if (@TypeOf(T) == u16) {
+                            break :brk varedLenBuf[0 ..bufSize];
+                        } else {
+                            var u16Buf: WPathBuffer = undefined;
+                            bufSize = std.unicode.utf8ToUtf16Le(&u16Buf, varedLenBuf[0 ..bufSize]) catch {
+                                return MaybeSlice(T).errnoSys(0, Syscall.Tag.getenv).?;
+                            };
+                            break :brk u16Buf[0 ..bufSize :0];
+                        }
+                    };
+                    // Zig's std.os.getenvW has logic to support keys like `=${resolvedDevice}`:
+                    // https://github.com/ziglang/zig/blob/7bd8b35a3dfe61e59ffea39d464e84fbcdead29a/lib/std/os.zig#L2126-L2130
+                    //
+                    // TODO: Enable test once spawnResult.stdout works on Windows.
+                    // test/js/node/path/resolve.test.js
+                    if (std.os.getenvW(key_w)) |r| {
+                        if (@TypeOf(T) == u16) {
+                            bufSize = r.len;
+                            @memcpy(varedLenBuf[0 ..bufSize], r);
+                        } else {
+                            // Reuse varedLenBuf because it's used for path.
+                            bufSize = std.unicode.utf16leToUtf8(varedLenBuf, r) catch {
+                                return MaybeSlice(T).errnoSys(0, Syscall.Tag.getcwd).?;
+                            };
```

Comment:
I'd like to learn more about when/how bun does things beyond the support of Zig. At the moment it seems inconsistent. We may use `bun.strings.convertUTF16...` but we call `std.os.getcwd` which calls `std.os.windows.GetCurrentDirectory` which uses `std.unicode.Utf16LeIterator` which `utf16leToUtf8` uses under the hood ([in addition to other optimizations](https://github.com/ziglang/zig/blob/0aff1f9ce51fd1b0df5cb5a6a06c5d6c744749fe/lib/std/unicode.zig#L860C8-L891)). It too also uses `simd` but uses [`std.simd`](https://github.com/ziglang/zig/blob/master/lib/std/simd.zig) instead of calling directly out to C like we do but __does not__ handle surrogates (https://github.com/ziglang/zig/issues/2565).

Reply:
😢 Currently our surrogate support in `convertUTF16ToUTF8InBuffer` is commented out and marked as TODO:

https://github.com/oven-sh/bun/blob/f77b217abf2f33fdea4a50298d07cd601a655b0d/src/string_immutable.zig#L5374-L5388

The surrogate safe option is `convertUTF16ToUTF8` and `convertUTF16ToUTF8Append`:
https://github.com/oven-sh/bun/blob/f77b217abf2f33fdea4a50298d07cd601a655b0d/src/string_immutable.zig#L1835-L1863

Reply:
I'll use `bun.strings.convertUTF16ToUTF8InBuffer` as a placeholder so once we add support we'll get it here too.

Reply:
Sounds good. I agree that we should look into all the places where we rely on zig stdlib unicode conversion.